### PR TITLE
Register lab2 labs in: `apps/lab2Entrypoints.ts`

### PR DESCRIPTION
**Pull request by snickell**
### The goal of this PR is to:
1. Raise lab2 entrypoint definitions to the top-level (parallel to webpackEntryPoints.js used by lab1)
2. Keep lab-specific configuration out of the `src/lab2` folder, making it framework-only
3. Discourage repeating, in new lab2 based labs, the non-lazy-loading pattern that now dominates (generates a single all-lab2-mega-bundle, doesn't scale well with adding new labs, bad network perf)
4. Better document how to make a lab2

### What this PR does
- Refactor where lab2s are registered, moving them to `app/lab2EntryPoints.ts` (from `app/src/lab2/views/LabViewsRenderer.tsx`). This encapsulates lab-specific definitions inside the lab's source folder, and reserves src/lab2 for just framework code, encouraging looser coupling between the lab2 framework and its instances.
- Extracts each lab's `Lab2EntryPoint` and moves it into the lab's source folder e.g. `apps/src/pythonlab/entrypoint.ts`
- Renames the `AppProperties` type to `Lab2EntryPoint` for consistency with filenames, and moves it from `LabViewsRenderer.tsx` => `types.ts`.
  - Derives the previously not-DRY AppName type automatically from the keys of `lab2EntryPoints`.
  - Renames AppProperties.lazyNode => Lab2EntryPoint.view, and AppProperties.node => Lab2EntryPoint.harcodedView, encouraging future lab2 authors of thinking of lazy-loading as "the default way to do it".
    - Additionally, to discourage following the now dominant pattern of NOT lazy loading anything, to use `Lab2EntryPoint.hardcodedView` you must set: `Lab2EntryPoint.view = OptionsToAvoid.UseHardcodedEntryPoint_WARNING_Bloats_Lab2_Bundle`. Using hardcoded views widely will result in every lab2-based lab having to load every other lab2-based lab's code to start, and is not a sustainable pattern for google school internet performance.
- Copies the [Lab2 Architecture doc's "How to Create a New Lab" section](https://docs.google.com/document/d/1a8fPxd4mvV7LdhxGOCuCzTs4cmNt0o18CwpcH2KeGBU/edit) into [apps/src/lab2/README.md](https://github.com/code-dot-org/code-dot-org/blob/9db9e9c0401de4a35b903c870542c47f7f622eb6/apps/src/lab2/README.md#how-to-create-a-new-lab), and updates it to reflect changes in this PR.
 
#### lab2EntryPoints.ts:
```
import {MusicEntryPoint} from '@cdo/apps/music/entrypoint';
import {PanelsEntryPoint} from '@cdo/apps/panels/entrypoint';
// ...

export const lab2EntryPoints: Record<string, Lab2EntryPoint> = {
  music: MusicEntryPoint,
  panels: PanelsEntryPoint,
  // ...
};
```

#### src/music/entrypoint.ts:
```
export const MusicEntryPoint: Lab2EntryPoint = {
  backgroundMode: true,
  theme: Theme.DARK,
  setupFunction: setUpBlocklyForMusicLab,
  view: OptionsToAvoid.UseHardcodedEntryPoint_WARNING_Bloats_Lab2_Bundle,
  hardcodedView: <MusicView />,
};
```

### TODO:
- [ ] update [Lab2 Architecture doc](https://docs.google.com/document/d/1a8fPxd4mvV7LdhxGOCuCzTs4cmNt0o18CwpcH2KeGBU/edit) to point to updated "How to Add a New Lab" instructions in `app/src/lab2/README.md`.
- [ ] verify that setting Lab2EntryPoint.hardcodedView as e.g. `Music` vs `<Music />` won't screw anything up with page transitions (it looks like it shouldn't.... but would like to verify locally or confirm tests would catch anything if I broke it)
- [ ] Ugh, even though lazy-loaded views are definitely worth it, they do make starting up a new lab more complicated. The biggest issue is all the type fiddling with having to use an index.js to work with dynamic import. Is there any way to simplify this???

### CODE REVIEW

File: apps/src/lab2/README.md [hunk 1]
```diff
@@ -21,9 +23,29 @@ This folder contains the code for **Lab2**, which is our set of shared component
...
 - An error boundary which catches errors, reports them, and displays an error message.
 
+## How to Create a New Lab
+
+As of 2024, all new labs should be based on Lab2. To create a new lab, you need to do the following:
+
+- Backend changes (ruby):
+  - Set up the new [level](https://github.com/code-dot-org/code-dot-org/pull/52538/files#diff-3c5dedf7c252ccdfebcad7b26ddc2b4e1d2cad5c5d4257943802c0af06fb7c83R1) and [game type](https://github.com/code-dot-org/code-dot-org/pull/52538/files#diff-ff86f86a12d97e95649d41a3da445d7e1fb3cdcf69ef4112e1c8d8a02fee7233R62), and add the new level to [levels_controller](https://github.com/code-dot-org/code-dot-org/pull/52538/files#diff-fb10c591e5aa0c82109ab9b5e3d5aac6f7ffdf5ebb2717de405afa7c661dd650R20). This is unchanged from how our previous labs were set up.
+  - Override [uses_lab2](https://github.com/code-dot-org/code-dot-org/pull/52538/files#diff-3c5dedf7c252ccdfebcad7b26ddc2b4e1d2cad5c5d4257943802c0af06fb7c83R48) in the new level file and return true.
+  - There are a few other minor backend changes required, see the linked PR.
```
Comment (molly-moen) [inline apps/src/lab2/README.md:34]: nit: I think this is referring to the pr that's linked in various places above, but when looking at the rendered markdown it won't be obvious what the "linked PR" is until you click on one of the above links. Mind adding a link at the top with some message like: this is modeled off of the following PR?

File: apps/src/lab2/README.md [hunk 2]
```diff
@@ -21,9 +23,29 @@ This folder contains the code for **Lab2**, which is our set of shared component
...
+  - Register your lab with lab2, these steps have changed since the aichat PR so we'll be looking at an example from pythonlab instead:
+    - Define a new entrypoint for your lab, see [apps/pythonlab/entrypoint.ts](https://github.com/code-dot-org/code-dot-org/pull/59794/files#diff-0dbe9f4f6eecfd5cb5e653b9d592bacbaaf6368602e32a1c357ce7c9067f0cdb)
+      - DO: use `view:` with a dynamic import() call with a webpack chunk name defined
+      - DO NOT: use `hardcodedView:` with a static import of your react view, even if you see examples doing this (why: this results in all lab2 code being loaded in a huge mega-bundle)
+    - Create an index.js file for your lab, see [apps/pythonlab/index.js](https://github.com/code-dot-org/code-dot-org/blob/b81b6b98f1ecc01972bf2ae43589b06c40e6183c/apps/src/pythonlab/index.js), this is required to make our typescript config work with dynamic imports.
+    - Add the new lab type to [apps/lab2EntryPoints.ts](https://github.com/code-dot-org/code-dot-org/blob/b81b6b98f1ecc01972bf2ae43589b06c40e6183c/apps/lab2EntryPoints.ts#L21)
+
+That’s it!
```
Comment (molly-moen) [inline apps/src/lab2/README.md:44]: 😅

File: apps/src/lab2/types.ts [hunk 3]
```diff
@@ -177,6 +184,58 @@ export interface VideoLevelData {
...
+   * @deprecated: using this option will result in hardcoding this lab into the
+   * downloaded bundle for ALL other lab2 labs, slowing down their loading and
+   * consuming excessive school internet bandwidth.
+   *
+   * See `pythonlab/entrypoint.tsx` for an example that doesn't use this option.
+   *
+   * Please only use this option if there's a good reason you can't lazy load
+   * your lab. With this option set, you must also specify `hardcodedEntryPoint`.
+   */
+  UseHardcodedView_WARNING_Bloats_Lab2_Bundle,
```
Comment (molly-moen) [inline apps/src/lab2/types.ts:198]: Thanks for the rename!

**Review by molly-moen — APPROVED** [review]
Thanks for doing this, this is much cleaner! I tested the lab2 showcase progression (`/s/allthethings/lessons/52/`) and things seem to be working as expected.

File: apps/src/lab2/types.ts (hunk 3 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/types.ts:198]: +1, very clear (and appropriately scary 😆)

File: apps/src/lab2/types.ts [hunk 4]
```diff
@@ -201,28 +260,7 @@ export type ProjectType =
...
-  | 'craft'
-  | 'studio'
-  | 'bounce'
-  | 'poetry'
-  | 'pythonlab'
-  | 'spritelab'
-  | 'standalone_video'
-  | 'panels'
-  | 'weblab2';
+export type AppName = keyof typeof lab2EntryPoints;
```
Comment (sanchitmalhotra126) [inline apps/src/lab2/types.ts:263]: 🎉

**Review by sanchitmalhotra126 — APPROVED** [review]
Looks great, thanks for doing this! Looking forward to moving more of these labs over to lazy-loading after this

File: apps/src/lab2/README.md (hunk 1 above)
Comment (snickell) [reply apps/src/lab2/README.md:34]: Good point, was definitely in markdown mind

File: apps/src/lab2/types.ts (hunk 3 above)
Comment (snickell) [reply apps/src/lab2/types.ts:198]: 👹

**Pull request merged**